### PR TITLE
[flang] Allow `VariableAssignBufferization` to handle `hlfir::ExprType`

### PR DESCRIPTION
**Pull request by ergawy**
Given the following input:
```fortran
1. subroutine ComputeDifferencesKernel
2.   implicit none
3.   integer :: i
4.   integer, dimension(1) :: a
5.
6.   do i = 1, 10
7.     a = [ i ]
8.   end do
9. end subroutine ComputeDifferencesKernel
```
Currently, the assignment in line 7 ends up as a call to the Fortran runtime function: `AAssign` since the corresponding `hlfir.assign` is not optimized away by `VariableAssignBufferization`. The reason this assignment is not optimized away is that `VariableAssignBufferization` does not match whenever the RHS of the assignment is a `hlfir.expr` value. However, this behavior is introduced only to prevent clashes between `VariableAssignBufferization` and `ElementalAssignBufferization` which optimizes away assignemnts that result from `hlfir.elemental` ops.

This patch relaxes that restriction by checking whether the RHS of an `hlfir.assign` is the result of `hlfir.elemental` or not. If not, we can safely proceed with `VariableAssignBufferization`.

Note that in the above example, we won't get a `hlfir.elemental` in the IR. We would get if we changed line 7 to something like:
```fortran
7.     a = [ i ] + b
```
In which case, `ElementalAssignBufferization` will kick in instead.

### CODE REVIEW

File: flang/test/HLFIR/opt-bufferization.fir
Comment (tblah) [inline flang/test/HLFIR/opt-bufferization.fir:801]: nit
```suggestion
// `hlfir.elemental`)
```

File: flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp
Comment (tblah) [inline flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:597]: nit (feel free to ignore if you disagree)
```suggestion
```

**Review by tblah — APPROVED** [review]
This seems like a good solution to me. Please wait for approval from @vzakhari too

**Comment by ergawy** [comment]
Thanks for the review @tblah. Ping for other review 🔔! Anyone objecting to merging this PR (maybe @vzakhari)?

**Comment by ergawy** [comment]
One more ping 🔔! Please let me know if there are any objections to merging the PR.

**Pull request merged**
